### DISCUSSION TITLE
Zeek plugins and cmake compatibility

### DISCUSSION THREAD

**Discussion by ckreibich** [zeek/zeek#2963] [General]
We're in the process of updating Zeek's cmake-level plugin API] (see [here](https://github.com/zeek/cmake/pull/60), [here](https://github.com/zeek/cmake/pull/64)), which we haven't done in a long time and brings up questions around how plugins transition to the new API and how we phase out the old one. We touched this a bit over in [this PR](https://github.com/zeek/package-template/pull/17) but since that's pretty peripheral I figured I'd start a discussion.

For context: the `zeek_plugin_start()`, ..., `zeek_plugin_end()` API for the plugins' CMakeLists.txt is going away, replaced by a more cmake-modern `zeek_add_plugin()` call that also cleans up a bunch of internal stuff, including Bro-era cruft. There remains a compatibility layer that provide the old API, even including the ancient `bro_plugin_*` API.

In the following I'm focusing on `zkg` packages that include plugins ("plugin-packages"), since that's how most folks encounter plugins.

- Backward compatibility of a package is always up to the author, not us. We have several plugin-packages that really go out of their way to build with old Zeeks. So far this has mainly affected the C++ code via layers of version-specific ifdefs, not so much the configure/cmake setup, but that strikes me as a detail: in the end it all codifies support for multiple versions. The CMakeLists.txt of plugins is generally straightforward — it's mostly calls to the Zeek plugin API, plus perhaps a few `find_package` ones. (There are exceptions, like including locally pre-compiled content — but the new API isn't really going near that, and that's expert terrain anyway.)
- People run old Zeeks. While we don't support anything prior to the current LTS cycle, there's plenty of Zeek 4, 3, and even 2 out there still (see survey results!). Not that we encourage that, but it's worth keeping in mind.
- Since packages can be versioned flexibly, one approach for package maintainers that enables support for the old and new APIs is to set aside separate release lines for different Zeek compatibility windows. Existing packages can transition to the new API cleanly with a new release, though the developer will need to maintain multiple release lines to integrate features/fixes for pre-6.0.0 compatibility. It's also harder for users to find the right version to use (but see below).
- Another approach is to combine both APIs from the same CMakeLists.txt, driven by availability of the new one (`if (COMMAND zeek_add_plugin)` or some such). That's a cmake-level analog to what we already to via ifdefs for C++. While not pretty, it buys the developer flexibility going forward and users don't need to pay special attention
- The above are up to the package maintainer, really. This leaves the question of what we do for new packages. If we update the template to only provide the new API, I think that's too harsh since for a good amount of time a new package won't build in popular environments (consider e.g. the window of time until Security Onion moves to new line of Zeek releases). On the other hand it also seems messy to codify the dual-API solution by default. I could see providing multiple plugin features in the template, say `plugin` and `plugin-5.0`, to give people a choice. (I think that's essentially your line of thought, @Neverlord). I don't think there's a risk of combinatorial explosion since we don't provide other plugin specializations.
- Well, there's one: Spicy — but afaik the template's Spicy functionality is independent of these APIs, but let me know if I'm off there.

Regarding deprecations & removal, I suggest we: 
- preserve the `zeek_plugin_*` and `bro_plugin_*` APIs through the 6.0 cycle,
-  deprecate both from 6.0 onward (we already do so for `bro_plugin_*`),
- remove both in 7.1, for a long deprecation cycle.

This reflects the fact that about 1/3 of plugins in the standard source still use `bro_plugin_*`.

A thought re. package versions: a package's `zkg.meta` can (and should) state a minimum required Zeek version. `zkg` could help figure out an appropriate package version for the local Zeek by trawling backward through the version history to suggest the highest that suits your Zeek version. This should come with a warning that you're not getting latest and greatest, as suggested by @0xxon in the past, in case the developers don't maintain multiple release lines.

Regarding `init-plugin` (i.e, bootstrapping pure plugins, not Zeek packages): this has always been more tightly coupled to the Zeek source tree (and doesn't get installed), so it seems fine for 6.0 to move that straight to the new API.

Please chime in here, particularly if you develop plugins. This can certainly go many ways.

**Comment by rsmmr** [comment]
Sounds generally good to me. Couple notes:

> * Another approach is to combine both APIs from the same CMakeLists.txt, driven by availability of the new one (`if (COMMAND zeek_add_plugin)` or some such).

That's strikes me as the most practical approach for most plugins. If we provide some boiler-plate to copy/paste, that should be easy for people to add if they care about compatibility across versions.

Might also be worth filing tickets with affected packages and include that boiler-plate code to work with all versions; ideally in advance of 6.0.

> On the other hand it also seems messy to codify the dual-API solution by default.

Not sure, it might actually be a good default at least for a while. Alternatively, we could provide features, as you say, but maybe we can phrase those in terms of versions: tell me which Zeek versions you want to support and I generate the right template for you.

> * Well, there's one: Spicy — but afaik the template's Spicy functionality is independent of these APIs, but let me know if I'm off there.

Yeah, should not affect Spicy analysers as long as they don't build a plugin on their own as well.

> * deprecate both from 6.0 onward (we already do so for `bro_plugin_*`),

One thing to keep in mind here is that it's users who will be seeing the deprecations without being able to do anything about it (other than ask the plugin maintainer to fix it). Not sure what that means, but that's always a bit unfortunate.
